packages/pathfinder/src/pathfinder.ts

```diff
@@ -0,0 +1,400 @@
+                        const key = this.makeKeyPair(asset);
+                        asset_dict[key] = true;
+                    });
+                    balances.map((asset: StellarSdk.Asset) => {
+                        const key = this.makeKeyPair(asset);
+                        if (!asset_dict[key] && !asset.isNative()) throw PathfinderInitializationError
+                    });
+                }
+                return true;
+            });
+    }
+
+    /****************************************
+     * Pathfinding Utilities
+     ****************************************/
+    public async getPath(params: {
+        sourceAsset: stellarAsset,
+        destinationAsset: stellarAsset,
+        destinationAmount: number,
+    }): Promise<IPathInfo> {
+        const { sourceAsset, destinationAsset, destinationAmount } = params;
+
+        // checks to make sure that both of the assets are supported in the database
+        if (!this.supportsAsset(sourceAsset)) {
+            throw new Error(`[getPath] Given an unsupported sourceAsset: ${sourceAsset}`);
+        } else if (!this.supportsAsset(destinationAsset)) {
+            throw new Error(`[getPath] Given an unsupported destinationAsset: ${destinationAsset}`);
+        }
+
+        // calculate the quanitized threshold amount
+        const bucket = await this.getBucket(destinationAsset, destinationAmount);
+
+        // check to see if the path info was cached
+        const cachedInfo = await this.loadPathFromCache(sourceAsset, destinationAsset, bucket);
+        if (cachedInfo) { return cachedInfo; }
+
+        // get the best paths for the destination asset and amount
+        const bestPaths = await this.getBestPaths(destinationAsset, destinationAmount);
+
+        // cache the results - no need to await this
+        this.savePathsToCache(destinationAsset, bucket, bestPaths);
+
+        // get the best path corresponding to the sourceAsset and return
+        const bestPath = bestPaths.find((pathInfo) => {
+            return pathInfo.from.issuerPublicKey === sourceAsset.issuerPublicKey &&
+                pathInfo.from.code == sourceAsset.code;
+        });
+
+        if (!bestPath) {
+            throw new Error("[getPath]: stellar account not configured correctly");
+        }
+
+        return bestPath;
+    }
+
+    /***************************************************************************
+     * Cache Logic
+     ***************************************************************************/
+
+    /**
+     * Queries the cache and returns a promise for the path information; if no path information
+     * exists, the promise resolves to null.
+     */
+    private async loadPathFromCache(
+        sourceAsset: stellarAsset,
+        destinationAsset: stellarAsset,
+        threshold: number,
+    ): Promise<IPathInfo | null> {
+        // todo provide the retrieval logic
+        const key = `${destinationAsset.code}_${destinationAsset.issuerPublicKey}`;
```

Comment:
this doesn't look like it is doing dependency injection. Additionally, the key formation logic should happen within the cache.

Reply:
i think i fixed